File: frontend/src/utils/validators.js
/**
 * Validation Utilities
 * Client-side validation functions
 */

/**
 * Validate email format
 */
export const validateEmail = (email) => {
  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  return emailRegex.test(email);
};

/**
 * Validate phone number (basic validation)
 */
export const validatePhone = (phone) => {
  if (!phone || !phone.trim()) return true; // Optional field
  
  // Remove common formatting characters
  const cleaned = phone.replace(/[\s\-\(\)\+]/g, '');
  
  // Check if it contains only digits and is between 10-15 characters
  return /^\d{10,15}$/.test(cleaned);
};

/**
 * Validate full name (must have at least first and last name)
 */
export const validateFullName = (name) => {
  if (!name || !name.trim()) return false;
  
  const parts = name.trim().split(/\s+/);
  return parts.length >= 2;
};

/**
 * Validate registration code format
 */
export const validateRegistrationCode = (code) => {
  // Format: VIP-XXXXXXXX (VIP- followed by 8 alphanumeric characters)
  return /^VIP-[A-Z0-9]{8}$/.test(code);
};

/**
 * Validate age (must be 18+)
 */
export const validateAge = (dateOfBirth) => {
  if (!dateOfBirth) return false;
  
  const today = new Date();
  const birthDate = new Date(dateOfBirth);
  
  let age = today.getFullYear() - birthDate.getFullYear();
  const monthDiff = today.getMonth() - birthDate.getMonth();
  
  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birthDate.getDate())) {
    age--;
  }
  
  return age >= 18;
};

/**
 * Validate date is in the future
 */
export const validateFutureDate = (date) => {
  const selectedDate = new Date(date);
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  
  return selectedDate > today;
};

/**
 * Format phone number to display format
 */
export const formatPhone = (phone) => {
  if (!phone) return '';
  
  // Remove all non-digit characters
  const digits = phone.replace(/\D/g, '');
  
  if (digits.length === 10) {
    return `(${digits.slice(0, 3)}) ${digits.slice(3, 6)}-${digits.slice(6)}`;
  } else if (digits.length === 11 && digits[0] === '1') {
    return `+1 (${digits.slice(1, 4)}) ${digits.slice(4, 7)}-${digits.slice(7)}`;
  }
  
  return phone; // Return as-is if format is unexpected
};

/**
 * Format date to readable string
 */
export const formatDate = (date) => {
  if (!date) return '';
  
  const d = new Date(date);
  return d.toLocaleDateString('en-US', {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  });
};

/**
 * Format date and time
 */
export const formatDateTime = (date) => {
  if (!date) return '';
  
  const d = new Date(date);
  return d.toLocaleString('en-US', {
    weekday: 'long',
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: 'numeric',
    minute: '2-digit',
    hour12: true,
  });
};

/**
 * Sanitize filename
 */
export const sanitizeFilename = (filename) => {
  return filename
    .replace(/[^\w\s\-\.]/g, '')
    .replace(/[\s]+/g, '_')
    .trim();
};

/**
 * Validate file type
 */
export const validateFileType = (file, allowedTypes) => {
  if (!file) return false;
  return allowedTypes.includes(file.type);
};

/**
 * Validate file size
 */
export const validateFileSize = (file, maxSizeInMB) => {
  if (!file) return false;
  const maxSizeInBytes = maxSizeInMB * 1024 * 1024;
  return file.size <= maxSizeInBytes;
};

/**
 * Get error message for field validation
 */
export const getValidationError = (field, value) => {
  switch (field) {
    case 'email':
      if (!value) return 'Email is required';
      if (!validateEmail(value)) return 'Please enter a valid email address';
      return null;
      
    case 'name':
      if (!value) return 'Name is required';
      if (!validateFullName(value)) return 'Please enter your full name (first and last name)';
      return null;
      
    case 'phone':
      if (value && !validatePhone(value)) return 'Please enter a valid phone number';
      return null;
      
    case 'age':
      if (!value) return 'Date of birth is required';
      if (!validateAge(value)) return 'You must be at least 18 years old';
      return null;
      
    default:
      return null;
  }
};

/**
 * Validate form data
 */
export const validateFormData = (formData, requiredFields) => {
  const errors = {};
  
  requiredFields.forEach((field) => {
    const error = getValidationError(field, formData[field]);
    if (error) {
      errors[field] = error;
    }
  });
  
  return {
    isValid: Object.keys(errors).length === 0,
    errors,
  };
};

/**
 * Truncate text
 */
export const truncateText = (text, maxLength = 100) => {
  if (!text || text.length <= maxLength) return text;
  return text.substring(0, maxLength) + '...';
};

/**
 * Capitalize first letter
 */
export const capitalizeFirst = (str) => {
  if (!str) return '';
  return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
};

/**
 * Generate initials from name
 */
export const getInitials = (name) => {
  if (!name) return '';
  
  const parts = name.trim().split(/\s+/);
  if (parts.length === 1) return parts[0].charAt(0).toUpperCase();
  
  return (parts[0].charAt(0) + parts[parts.length - 1].charAt(0)).toUpperCase();
};

File: backend/app/utils/validators.py
"""
Validation Utilities
Helper functions for validation and code generation
"""

import re
import secrets
import string
from datetime import datetime
from typing import Optional


def validate_email(email: str) -> bool:
    """
    Validate email format
    
    Args:
        email: Email address to validate
        
    Returns:
        bool: True if valid, False otherwise
    """
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def validate_phone(phone: str) -> bool:
    """
    Validate phone number format (basic validation)
    Accepts various formats: (123) 456-7890, 123-456-7890, 1234567890
    
    Args:
        phone: Phone number to validate
        
    Returns:
        bool: True if valid, False otherwise
    """
    if not phone or not phone.strip():
        return True  # Optional field
    
    # Remove common formatting characters
    cleaned = re.sub(r'[\s\-\(\)\+]', '', phone)
    
    # Check if it contains only digits and is between 10-15 characters
    return bool(re.match(r'^\d{10,15}$', cleaned))


def validate_registration_code(code: str) -> bool:
    """
    Validate registration code format
    
    Args:
        code: Registration code to validate
        
    Returns:
        bool: True if valid format
    """
    # Format: VIP-XXXXXXXX (VIP- followed by 8 alphanumeric characters)
    pattern = r'^VIP-[A-Z0-9]{8}$'
    return bool(re.match(pattern, code))


def generate_registration_code() -> str:
    """
    Generate a unique registration code for a fan
    Format: VIP-XXXXXXXX
    
    Returns:
        str: Registration code
    """
    # Generate 8 random uppercase alphanumeric characters
    random_part = ''.join(secrets.choice(string.ascii_uppercase + string.digits) for _ in range(8))
    return f"VIP-{random_part}"


def generate_ticket_id(fan_id: int, tour_id: int) -> str:
    """
    Generate a unique ticket ID
    Format: TKT-{TIMESTAMP}-{FAN_ID}-{TOUR_ID}-{RANDOM}
    
    Args:
        fan_id: Fan ID
        tour_id: Tour ID
        
    Returns:
        str: Unique ticket ID
    """
    timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
    random_suffix = ''.join(secrets.choice(string.ascii_uppercase + string.digits) for _ in range(4))
    return f"TKT-{timestamp}-{fan_id}-{tour_id}-{random_suffix}"


def sanitize_filename(filename: str) -> str:
    """
    Sanitize filename to remove unsafe characters
    
    Args:
        filename: Original filename
        
    Returns:
        str: Sanitized filename
    """
    # Remove or replace unsafe characters
    filename = re.sub(r'[^\w\s\-\.]', '', filename)
    filename = re.sub(r'[\s]+', '_', filename)
    return filename.strip('._')


def validate_age(date_of_birth: datetime, min_age: int = 18) -> bool:
    """
    Validate that a person is at least min_age years old
    
    Args:
        date_of_birth: Date of birth
        min_age: Minimum required age (default: 18)
        
    Returns:
        bool: True if person is old enough
    """
    from datetime import date
    
    if isinstance(date_of_birth, datetime):
        dob = date_of_birth.date()
    else:
        dob = date_of_birth
    
    today = date.today()
    age = today.year - dob.year - ((today.month, today.day) < (dob.month, dob.day))
    return age >= min_age


def format_phone(phone: str) -> str:
    """
    Format phone number to a standard format
    Example: 1234567890 -> (123) 456-7890
    
    Args:
        phone: Phone number
        
    Returns:
        str: Formatted phone number
    """
    # Remove all non-digit characters
    digits = re.sub(r'\D', '', phone)
    
    if len(digits) == 10:
        return f"({digits[:3]}) {digits[3:6]}-{digits[6:]}"
    elif len(digits) == 11 and digits[0] == '1':
        return f"+1 ({digits[1:4]}) {digits[4:7]}-{digits[7:]}"
    else:
        return phone  # Return as-is if format is unexpected


def truncate_string(text: str, max_length: int = 100, suffix: str = "...") -> str:
    """
    Truncate a string to a maximum length
    
    Args:
        text: Text to truncate
        max_length: Maximum length
        suffix: Suffix to add if truncated
        
    Returns:
        str: Truncated text
    """
    if len(text) <= max_length:
        return text
    return text[:max_length - len(suffix)] + suffix